src/cargo/util/auth.rs

```diff
@@ -273,37 +330,168 @@ pub fn auth_token(config: &Config, sid: &SourceId, login_url: Option<&Url>) -> C
+                        | Mutation::Unyank { name, .. }
+                        | Mutation::Owners { name, .. } => *name,
+                    })
+                }),
+                vers: mutation.and_then(|m| {
+                    Some(match m {
+                        Mutation::PrePublish | Mutation::Owners { .. } => return None,
+                        Mutation::Publish { vers, .. }
+                        | Mutation::Yank { vers, .. }
+                        | Mutation::Unyank { vers, .. } => *vers,
+                    })
+                }),
+                cksum: mutation.and_then(|m| {
+                    Some(match m {
+                        Mutation::PrePublish
+                        | Mutation::Yank { .. }
+                        | Mutation::Unyank { .. }
+                        | Mutation::Owners { .. } => return None,
+                        Mutation::Publish { cksum, .. } => *cksum,
+                    })
+                }),
+                challenge: None, // todo: PASETO with challenges
+                v: None,
+            };
+            let footer = Footer {
+                url: &sid.url().to_string(),
+                kip,
+            };
+
+            pasetors::version3::PublicToken::sign(
+                &secret,
+                serde_json::to_string(&message)
+                    .expect("cannot serialize")
+                    .as_bytes(),
+                Some(
+                    serde_json::to_string(&footer)
+                        .expect("cannot serialize")
+                        .as_bytes(),
+                ),
+                None,
+            )?
+        }
     };
 
-    cache.insert(url.clone(), token.clone());
+    if mutation.is_none() {
+        cache.insert(
+            url.clone(),
+            CredentialCacheValue {
+                from_commandline: false,
+                token_value: token.to_string(),
+            },
+        );
+    }
     Ok(Some(token))
 }
 
+pub enum Mutation<'a> {
+    PrePublish,
+    Publish {
+        name: &'a str,
+        vers: &'a str,
+        cksum: &'a str,
+    },
+    Yank {
+        name: &'a str,
+        vers: &'a str,
+    },
+    Unyank {
+        name: &'a str,
+        vers: &'a str,
+    },
+    Owners {
+        name: &'a str,
+    },
+}
+
+#[derive(serde::Serialize)]
+struct Message<'a> {
+    iat: &'a str,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    sub: Option<&'a str>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    mutation: Option<&'a str>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    name: Option<&'a str>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    vers: Option<&'a str>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    cksum: Option<&'a str>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    challenge: Option<&'a str>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    v: Option<u8>,
```

Comment:
The `v` field does not seem to be used. I'm inferring this would only be set in the future (to something like `Some(2)`) if there is a breaking change? Can some documentation about this be added here?